File: 0212-word-search-ii/0212-word-search-ii.cpp
class Solution {
public:
    struct TrieNode {
        TrieNode *child[26];
        string word;
        TrieNode() : word("") {
            for (auto &a : child) a = nullptr;
        }
    };

    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        TrieNode *root = buildTrie(words);
        vector<string> res;
        for(int i = 0; i < board.size(); i++)
        {
            for (int j = 0; j< board[0].size(); j++)
            {
                dfs(board, i, j, root, res);
            }
        }
        return res;
    }

    private:
    void dfs(vector<vector<char>>& board, int i, int j, TrieNode* p, vector<string>& res)
    {
        char c = board[i][j];
        if (c == '#' || p->child[c - 'a'] == nullptr) return;
        p = p->child[c - 'a'];
        if (p->word != "")
        {
            res.push_back(p->word);
            p->word = "";
        }
        board[i][j] = '#';
        if (i > 0) dfs(board, i - 1, j, p, res);
        if (j > 0) dfs(board, i, j - 1, p, res);
        if (i < board.size() - 1) dfs(board, i + 1, j, p, res);
        if (j < board[0].size() - 1) dfs(board, i, j + 1, p, res);
        board[i][j] = c;
    }
    TrieNode* buildTrie(vector<string>& words) {
        TrieNode *root = new TrieNode();
        for(string &w : words) {
            TrieNode *p = root;
            for (char c : w) {
                int i = c - 'a';
                if(p->child[i] == nullptr) p->child[i] = new TrieNode();
                p = p->child[i];
            }
            p->word = w;
        }
        return root;
    }
};

File: 0212-word-search-ii/0212-word-search-ii.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False
        self.word = None

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        
        # Build the Trie
        for word in words:
            node = root
            for char in word:
                if char not in node.children:
                    node.children[char] = TrieNode()
                node = node.children[char]
            node.is_end_of_word = True
            node.word = word
        
        def dfs(x, y, node):
            char = board[x][y]
            if char not in node.children:
                return
            node = node.children[char]
            if node.is_end_of_word:
                found_words.add(node.word)
            
            board[x][y] = '#'  # Mark the cell as visited
            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:  # Up, down, left, right
                nx, ny = x + dx, y + dy
                if 0 <= nx < m and 0 <= ny < n and board[nx][ny] != '#':
                    dfs(nx, ny, node)
            board[x][y] = char  # Unmark the cell
            
        m, n = len(board), len(board[0])
        found_words = set()
        
        # Start DFS from each cell
        for i in range(m):
            for j in range(n):
                dfs(i, j, root)
        
        return list(found_words)


